### DISCUSSION TITLE
How to broadcast an argument when using `nn.scan()`?

### DISCUSSION THREAD

**Discussion by andsteing** [google/flax#2127] [Q&A] [answered]
(Originally asked by Nanxin Chen)

I've read that "scanning happens over all arguments starting from the third argument" in the [`nn.scan()` docs](https://flax.readthedocs.io/en/latest/_autosummary/flax.linen.scan.html)

How can I make that some arguments are broadcast instead?

**Comment by andsteing** [comment, marked as answer]
(Originally answered by @levskaya )

You can specify an axis for every argument individually and use `nn.broadcast` to broadcast an argument.

In below example, the first argument `x` is scanned over axis 0, and the second argument `t` is broadcast in the scan.

```python
class Model(nn.Module):
  @nn.compact
  def __call__(self, x, t):
    LayerScanned = nn.scan(Layer,
                           variable_broadcast='params',
                           split_rngs={'params': False},
                           in_axes=(0, nn.broadcast),
                           out_axes=1)
    carry = jnp.zeros_like(x)
    carry, x = LayerScanned()(carry, x, t)
    return x
```

<details>
<summary>Full example</summary>

```python
import flax.linen as nn
import jax
import jax.numpy as jnp

class Layer(nn.Module):
  @nn.compact
  def __call__(self, c, x, t):
    x = nn.Dense(len(x))(x)
    x = jax.nn.softmax(jnp.exp(t) * x)
    c = c + x
    return c, x

class Model(nn.Module):
  @nn.compact
  def __call__(self, x, t):
    LayerScanned = nn.scan(Layer,
                           variable_broadcast='params',
                           split_rngs={'params': False},
                           in_axes=(0, nn.broadcast),
                           out_axes=1)
    carry = jnp.zeros_like(x)
    carry, x = LayerScanned()(carry, x, t)
    return x


x = jnp.zeros([3, 2])
model = Model()
vs = model.init(jax.random.PRNGKey(0), x, 1)

jax.tree_map(jnp.shape, vs)
```

output:

```
FrozenDict({
    params: {
        ScanLayer_0: {
            Dense_0: {
                bias: (2,),
                kernel: (2, 2),
            },
        },
    },
})
```
</details>

**Reply by cgarciae** [reply]
Similar to `jax.vmap` does `in_axes` in `nn.scan` also interpret `None` as `broadcast`?
